botocore/utils.py

```diff
@@ -794,5 +799,102 @@ def _switch_hosts(request, new_endpoint, use_new_scheme=True):
         ''
     )
     final_endpoint = urlunsplit(final_endpoint_components)
-    logger.debug('Updating URI from %s to %s' % (request.url, final_endpoint))
-    request.url = final_endpoint
+    logger.debug('Updating URI from %s to %s' % (
+        original_endpoint, final_endpoint))
+    return final_endpoint
+
+
+class S3RegionRedirector(object):
+    def __init__(self, endpoint_bridge):
+        self._endpoint_resolver = endpoint_bridge
+        self._cache = {}
+
+    def register(self, emitter):
+        emitter.register('needs-retry.s3', self.redirect_from_error)
+        emitter.register('before-call.s3', self.set_request_url)
+        emitter.register('before-parameter-build.s3', self.redirect_from_cache)
+
+    def redirect_from_error(self, request_dict, response, context, **kwargs):
+        """
+        An S3 request sent to the wrong region will return an error that
+        contains the endpoint the request should be sent to. This handler
+        will add the redirect information to the signing context and then
+        redirect the request.
+        """
+        error = response[1].get('Error', {})
+        if error.get('Code') != 'PermanentRedirect':
+            return
+
+        bucket = error['Bucket']
+        client_region = context.get('client_region')
+        new_region = self.get_region_from_response(response)
+
+        if new_region is None:
+            logger.debug(
+                "S3 client configured for region %s but the bucket %s is not "
+                "in that region and the proper region could not be "
+                "automatically determined." % (client_region, bucket))
+            return
+
+        logger.debug(
+            "S3 client configured for region %s but the bucket %s is in region"
+            " %s; Please configure the proper region to avoid multiple "
+            "unnecessary redirects and signing attempts." % (
+                client_region, bucket, new_region))
+
+        endpoint = self._endpoint_resolver.resolve('s3', new_region)
+        endpoint = endpoint['endpoint_url']
+
+        if context is None:
+            context = {}
+
+        signing_context = {
+            'region': new_region,
+            'bucket': bucket,
+            'endpoint': endpoint
+        }
+        context['signing'] = signing_context
+
+        self._cache[bucket] = signing_context
```

Comment:
Do you think we should strap a lock around modifying and grabbing from the cache? If there are multiple threads sharing a client is there a chance that we hit an issue from the fact that the threads are sharing state of the cache?